### ISSUE TITLE
Add alias support to the engine

### ISSUE THREAD

**Issue by Frassle** [pulumi/pulumi#9734]
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

Currently the engine supports a list of URNs to be checked as aliases when registering a resource. Every SDK however supports a more user friendly alias object where the user can just give parts of the old URN and have the rest filled in by default (e.g. same stack, project, parent, and type but a different name). Each SDK also needs logic to resolve all aliases with consideration of any parent resource having aliases. This is both for the obvious case where a parent has changed type (if parent was type A and is now B but correctly has an alias that it was type A, then a child resource needs aliases added to consider that as well), but also for the case where we expect that if a childs name has its parents name as a prefix and its parents name has changed that should be an alias as well (e.g. A child resource called "foo-child" with a parent called "foo" that has an alias that it used to be called "bar" should add an alias to the child that it was called "bar-child" as well).

As we introduce new SDKs all this alias logic needs duplicating to each of them. And the YAML SDK for example does not yet fully support aliases (https://github.com/pulumi/pulumi-yaml/issues/156).

This ticket is to track the work of adding a new array to the RegisterResource protocol allowing SDKs to send "smart aliases" which the engine will then resolve to URNs, plus work to make the engine handle all parent/child alias resolution (regardless of if "smart aliases" are sent or not).

**Comment by Frassle** [comment]
Re-opened as we had to revert the first version of this.

When re-adding consider:
A) Ensuring we don't generate and store duplicate alias URNS
B) Looking into only saving the alias used per parent rather than all the parents aliases
C) Any other speed/space improvements we can make here.

**Comment by t0yv0** [comment]
Additional motivation: this could fix alias explosion #10010 kind of issue.

**Comment by Frassle** [comment]
So having had time to think about this some more I think we can re-add this to the engine in a way that solves the alias explosion completely.
The important point to note is that while our parent could have many aliases it would have only have used one. So while the current design looks like:
```
for _, parentAlias := range this.parent.Aliases {
	aliases = append(aliases, inheritedChildAlias(this.type, this.name, this.parent.Name(), parentAlias))
	for _, alias := range this.aliases {
		childAlias := collapseAliasToUrn(goal, alias)
		aliasedChildType := childAlias.Type()
		aliasedChildName := childAlias.Name()
		inheritedAlias := inheritedChildAlias(aliasedChildType, aliasedChildName, this.Parent.Name(), parentAlias)
		aliases = append(aliases, inheritedAlias)
	}
}
```

We don't need to loop over _all_ aliases that the parent declared, the only alias matters is the one that the parent actually used, which the engine does track, something like:
```
if this.Parent.Alias {
	aliases = append(aliases, inheritedChildAlias(this.type, this.name, this.parent.Name(), this.Parent.Alias))
	for _, alias := range this.aliases {
		childAlias := collapseAliasToUrn(goal, alias)
		aliasedChildType := childAlias.Type()
		aliasedChildName := childAlias.Name()
		inheritedAlias := inheritedChildAlias(aliasedChildType, aliasedChildName, this.Parent.Name(), parentAlias)
		aliases = append(aliases, inheritedAlias)
	}
}
```
This means there's no multi-level multiplication of aliases as we go down parent chains. Each level just needs to check one alias of it's direct parent, and then the aliases for itself.
